parquet-column/src/main/java/org/apache/parquet/io/MessageColumnIO.java

```diff
@@ -315,28 +338,50 @@ private void setRepetitionLevel() {
     @Override
     public void startGroup() {
       if (DEBUG) log("startGroup()");
+      GroupColumnIO group = (GroupColumnIO) currentColumnIO;
+      // current group is not null, need to flush all the nulls that were cached before
+      if (hasNullCache(group))
+        flushCachedNulls(group);
 
-      ++ currentLevel;
+      ++currentLevel;
       r[currentLevel] = r[currentLevel - 1];
 
-      int fieldsCount = ((GroupColumnIO)currentColumnIO).getChildrenCount();
+      int fieldsCount = ((GroupColumnIO) currentColumnIO).getChildrenCount();
       fieldsWritten[currentLevel].reset(fieldsCount);
       if (DEBUG) printState();
     }
 
+    private boolean hasNullCache(GroupColumnIO group) {
+      IntArrayList nulls = groupNullCache.get(group);
+      return nulls != null && !nulls.isEmpty();
+    }
+
+
+    private void flushCachedNulls(GroupColumnIO group) {
```

Comment:
you could just put the if(hasNullCache(group)) part in this method and make startGroup() just always call flushCachedNulls()